### ISSUE TITLE
KMS keys do not work in S3Mock

### ISSUE THREAD

**Issue by greygoo5e** [adobe/S3Mock#702]
Hello, 

(Hopefully this is the correct forum for this)

I was able to make some progress setting up your docker container for the simulated S3 bucket. I was able to connect using the aws s3 command line utility with a redirect to the localhost:9191

I configured aws with:
`aws configure
access key=47110815
secret_access_key=c4353c4c-3318-460a-bdcc-b0a57bd8d9d8
region='us-east-1'`

Running the S3mock with 
`docker run --env initialBuckets="bucketa, bucketb" --env root=/files --env  validKmsKeys=arn:aws:kms:us-east-1:47110815:key/c51fdeea-f623-4a2b-90b5-15d72963cf9d,arn:aws:kms:us-east-1:47110815:key/c4353c4c-3318-460a-bdcc-b0a57bd8d9d8 -p 9090:9090 -p 9191:9191 -t adobe/s3mock`

Trying to connect with:
`aws s3 cp ./localfile.txt s3://bucketa/ --endpoint-url=https://localhost:9191 --sse=aws:kms --no-verify-ssl --sse-kms-key-id=47110815`

I'm using the --no-verify-ssl flag to get around the certificate failure on the s3mock connection.

I receive an error saying the key 47110815 does not exist. (I also tried passing the secret key here as the id with no luck)

I've tried a few variations of the cp and the one that does work doesn't require specifying the kms keys in order to write a file to the bucket:

`aws s3 cp ./localfile.txt s3://bucketa/ --endpoint-url=https://localhost:9191 --no-verify-ssl `

I'm currently using the command line just to ensure I've setup the s3mock correctly. According to your documentation is that if the KMS keys are not specified correctly the connect should return with a 400 error.

Can you provide a bit of guidance regarding the KMS keys and what I'm doing wrong?

Thanks!

**Comment by afranken** [comment]
@greygoo5e which version of the `aws-cli` are you using? And which version of the S3Mock?

When trying to reproduce this, I noticed that SSL connections with the `aws-cli` are not working with current versions of S3Mock at all, see #706 
I will try to look at a few other issues as well and then release a new version in the next days.

**Comment by afranken** [comment]
Should add, after fixing #706, connection with HTTPS and KMS keys works without problems for me.

**Comment by greygoo5e** [comment]
"Id": "sha256:ab218533a8ce00866c979166af9f508f1043a3fcc0e26ccba411ddddb5ae97b9",
       "RepoTags": [
            "adobe/s3mock:latest"

**aws-cli/2.7.16**

Without the --no-verify-ssl flag I get this stack trace: 

`[javax.net](http://javax.net/).ssl.SSLHandshakeException: Received fatal alert: unknown_ca
    at java.base/sun.security.ssl.Alert.createSSLException(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.Alert.createSSLException(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.Alert$AlertConsumer.consume(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.TransportContext.dispatch(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.SSLTransport.decode(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.SSLEngineImpl.decode(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.SSLEngineImpl.readRecord(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.SSLEngineImpl.unwrap(Unknown Source) ~[na:na]
    at java.base/sun.security.ssl.SSLEngineImpl.unwrap(Unknown Source) ~[na:na]
    at java.base/[javax.net](http://javax.net/).ssl.SSLEngine.unwrap(Unknown Source) ~[na:na]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).ssl.SslConnection.unwrap(SslConnection.java:429) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).ssl.SslConnection$DecryptedEndPoint.fill(SslConnection.java:718) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.server.HttpConnection.fillRequestBuffer(HttpConnection.java:352) ~[jetty-server-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260) ~[jetty-server-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).FillInterest.fillable(FillInterest.java:105) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:555) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).ssl.SslConnection.onFillable(SslConnection.java:410) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).ssl.SslConnection$2.succeeded(SslConnection.java:164) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).FillInterest.fillable(FillInterest.java:105) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at [org.eclipse.jetty.io](http://org.eclipse.jetty.io/).ChannelEndPoint$1.run(ChannelEndPoint.java:104) ~[jetty-io-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) ~[jetty-util-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) ~[jetty-util-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) ~[jetty-util-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) ~[jetty-util-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409) ~[jetty-util-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) ~[jetty-util-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) ~[jetty-util-9.4.46.v20220331.jar!/:9.4.46.v20220331]
    at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]`


Then if I use the --no-verify-ssl flag I dont see a stack trace, but it allows me to upload a file and the container doesn't validate my keys.

Hopefully this is the same as the #706 issue though. Thanks!

**Comment by afranken** [comment]
@greygoo5e I tested again, read AWS docs again, this never worked properly.

after #714 is merged and released, start S3Mock with:

```
docker run -p 9090:9090 -p 9191:9191 -e validKmsKeys="arn:aws:kms:us-east-1:1234567890:key/valid-test-key-ref" -e initialBuckets="bucket-a, bucket-b" -e LOGGING_LEVEL_ROOT=DEBUG adobe/s3mock:latest
```

and then use aws-cli like this:

`aws s3 cp ./test.json s3://bucket-a/ --endpoint-url=https://localhost:9191 --no-verify-ssl --sse aws:kms --sse-kms-key-id valid-test-key-id`

KMS keys have nothing to do with session keys, which we both tried to use in our first try...
aws-cli cp docs:
https://docs.aws.amazon.com/cli/latest/reference/s3/cp.html#options

Please note that S3Mock does not actually support KMS encryption, it will just reject requests made with KMS key IDs that were not configured during startup (in this aws-cli example `--sse aws:kms --sse-kms-key-id INVALID-test-key-id` )

*KMS also does not have anything to do with SSL*, so it does not matter whether you use the HTTP or the HTTPS endpoint.


HTTP examples:

valid key ID:
```
aws s3 cp ./test.json s3://bucket-a/ --endpoint-url=http://localhost:9090 --sse aws:kms --sse-kms-key-id valid-test-key-id

upload: ./test.json to s3://bucket-a/test.json
```

INVALID key ID:
```
aws s3 cp ./test.json s3://bucket-a/ --endpoint-url=http://localhost:9090 --sse aws:kms --sse-kms-key-id INVALID-test-key-id

upload failed: ./test.json to s3://bucket-a/test.json An error occurred (KMS.NotFoundException) when calling the PutObject operation: Key ID INVALID-test-key-id does not exist!
```

**Comment by afranken** [comment]
@greygoo5e also, looking at your first example again, you configured the KMS keys like this:
` validKmsKeys=arn:aws:kms:us-east-1:47110815:key/c51fdeea-f623-4a2b-90b5-15d72963cf9d,arn:aws:kms:us-east-1:47110815:key/c4353c4c-3318-460a-bdcc-b0a57bd8d9d8`

and you expected the key ID to be `47110815`. The ARN format is: `arn:aws:kms:region:acct-id:key/key-id`, so `47110815` is actually the AWS account ID.

In your case, the KMS key IDs to pass to the AWS SDK / CLI would be `c51fdeea-f623-4a2b-90b5-15d72963cf9d` and `c4353c4c-3318-460a-bdcc-b0a57bd8d9d8` :)

**Comment by greygoo5e** [comment]
Hello Arne, 

Thanks for clarifying what the KMS keys are used for. I don't have a live S3 bucket so I'm kinda flying a bit blind here, with just your container for validation.

I totally expected what you describe in the post from the 23rd to be the case, I was assuming that when I used the keys on the unencrypted connection and was able to do things like `ls` and `cp` that maybe the protections with the keys only applied to the https connection. Glad to again have some clarification here. 

I'll give it a shot with the newest version as it looks like its already merged. Do you guys release a new docker container with every new master branch 'release/update'?

I did actually try your latest example regarding the correct key-id using keys `c51fdeea-f623-4a2b-90b5-15d72963cf9d` and `c4353c4c-3318-460a-bdcc-b0a57bd8d9d8`  which I assumed were the secret keys. I guess it would have been good practice to look up the actual S3 spec :)

Thanks!

**Comment by afranken** [comment]
The releases are done manually from time to time.
I still have one open issue left to investigate and expect a release in the next few days. :)

**Comment by afranken** [comment]
Just released `2.4.14` with the fix. :)
